File: S4-Clients/Java-client/src/main/java/com/ontotext/s4/service/Annotation.java
/*
 * S4 Java client library
 * Copyright 2016 Ontotext AD
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.ontotext.s4.service;

import java.util.HashMap;
import java.util.Map;

import com.fasterxml.jackson.annotation.JsonAnySetter;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;

/**
 * Representation of a single annotation from an annotated document returned by the S4 online API.
 */
public class Annotation {

    /**
     * Start offset of the annotation (a zero-based index into the
     * containing document's {@link AnnotatedDocument#text plain text}).
     */
    private long startOffset;

    /**
     * End offset of the annotation (a zero-based index into the
     * containing document's {@link AnnotatedDocument#text plain text}).
     */
    private long endOffset;

    /**
     * The annotation's features.
     */
    private Map<String, Object> features = new HashMap<>();

    public long getStartOffset() {
        return this.startOffset;
    }
    public void setStartOffset(long startOffset) {
        this.startOffset = startOffset;
    }

    public long getEndOffset() {
        return this.endOffset;
    }
    public void setEndOffset(long endOffset) {
        this.endOffset = endOffset;
    }

    public Map<String, Object> getFeatures() {
        return this.features;
    }
    public void setFeatures(Map<String, Object> features) {
        this.features = features;
    }

    /**
     * Used by Jackson to decode the "indices" property of the JSON response into start and end offsets.
     *
     * @param indices a two-element array containing the start and end offsets, in that order.
     */
    @JsonCreator
    public Annotation(@JsonProperty("indices") long[] indices) {
        startOffset = indices[0];
        endOffset = indices[1];
    }

    /**
     * Catch-all setter used by Jackson to gather other properties from the JSON response into the {@link #features} map.
     */
    @JsonAnySetter
    public void addFeature(String name, Object value) {
        features.put(name, value);
    }

}

File: S4-Clients/C#-client/S4ClientApi/Model/Annotation/Annotation.cs
/*
 * S4 C# client library
 * Copyright 2016 Ontotext AD
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Ontotext.S4.common;

namespace Ontotext.S4.service
{
    public class Annotation
    {
        private long StartOffset;
        private long EndOffset;
        private Dictionary<String, Object> Features = new Dictionary<String, Object>();

        public Annotation()
        {

        }

        public Annotation(long startOffset, long endOffset, Dictionary<String, Object> features)
        {
            this.startOffset = startOffset;
            this.endOffset = endOffset;
            this.features = features;
        }

        /// <summary>
        /// Start offset of the annotation (a zero-based index into the
        /// containing document's {@link AnnotatedDocument#text plain text}).
        /// </summary>
        public long startOffset
        {
            get { return StartOffset; }
            set { StartOffset = value; }
        }

        /// <summary>
        /// End offset of the annotation (a zero-based index into the
        /// containing document's {@link AnnotatedDocument#text plain text}).
        /// </summary>
        public long endOffset
        {
            get { return EndOffset; }
            set { EndOffset = value; }
        }

        /// <summary>
        /// The annotation's features.
        /// </summary>
        public Dictionary<String, Object> features
        {
            get { return Features; }
            set { Features = value; }
        }

        /// <summary>
        /// Used by Jackson to decode the "indices" property of the JSON response into start and end offsets.
        /// </summary>
        /// <param name="indices">indices a two-element array containing the start and end
        ///          offsets, in that order.</param>
        public Annotation(long[] indices)
        {
            StartOffset = indices[0];
            EndOffset = indices[1];
        }

        /// <summary>
        /// Catch-all setter used by Jackson to gather other properties from the JSON response into the {@link #features} map.
        /// </summary>
        /// <param name="name"></param>
        /// <param name="value"></param>
        public void addFeature(String name, Object value)
        {
            Features.Add(name, value);
        }
    }
}
